### DISCUSSION TITLE
How to use DI for Aspects

### DISCUSSION THREAD

**Discussion by martonx** [metalama/Metalama#14] [Q&A] [answered]
Hi,

I want to use Aspects for logging in an Asp.Net Core project.
I would need Logger<MyAspect> or HttpContextAccessor from DI. I have a specially configured Serilog for Logger. Reaching HttpContextAccessor is another quite trivial requirement I think.
How can I get these through DI? I know I should add these for every of my classes where I want to use this Aspect and use meta.This.logger, meta.This.httpContextAccessor in my aspect, but no way.

Could you suggest some good approach for getting DI in Aspects?

**Comment by prochan2** [comment]
Hello,

Here is an example of how to log all methods of all types which provide the logger as a field:

```C#
internal class LogFabric : ProjectFabric
{
    public override void AmendProject(IProjectAmender amender)
    {
        amender.With(p => p.Types.Where(
            t => t.Fields.Any(
                f => f.Name == "_logger" 
                && f.Type is INamedType nt 
                && nt.IsSubclassOf(t.Compilation.TypeFactory.GetTypeByReflectionName("Microsoft.Extensions.Logging.ILogger"))))
        .SelectMany(t => t.Methods)).AddAspect<LogAttribute>();
    }
}

public sealed class LogAttribute : OverrideMethodAspect
{
    public override dynamic? OverrideMethod()
    {
        meta.This._logger.LogInformation($"The method {meta.Target.Method.DeclaringType.FullName}.{meta.Target.Method.Name} is being executed.");
        return meta.Proceed();
    }
}
```

**Comment by martonx** [comment]
@prochan2 Thank you for your answer!
Yes, but what about if those methods, classes don't have a logger field yet? Now, we have to add logger field, and use DI to create logger in EVERY class, where we want to log the methods. Instead of this, the perfect solution could be: MetaLama could have a logger inside the Aspect, and it should use that logger.
And logging was just one example, what if we need HttpContextAccessor. Should we add it to all of our classes with ILogger, just to use it in LogAttribute?
We should add ILogger and HttpContextAccessor just for the aspect (through an annotation or constructor for example), and compiler should add these dependencies to the compiled code.

**Reply by gfraiteur** [reply, marked as answer]
You can introduce the `_logger` field from the aspect like this:

```cs
[Introduce]
private ILogger _logger;
```

There are many ways to initialize the logger field. See https://doc.metalama.net/aspects/advising/initializers.

However there is currently no way to add a parameter to a constructor.

**Comment by prochan2** [comment]
I see. There are multiple issues here.

There cannot be a logger "inside the aspect", because there's no run-time instance of an aspect in Metalama.

We could inject the `_logger` field into each target class which doesn't have one, but the problem is to initialize it. With just ASP.NET DI, we'd need to add a parameter to the constructor of the target class, which is only possible with Metalama SDK at the moment. (Like in https://github.com/postsharp/Metalama.Open.AutoCancellationToken .)

You could also keep a static reference to `ILoggerFactory` and use it to initialize the field using initializer.

**Comment by martonx** [comment]
Thank you for the great ideas! Anyway none of them are the real good solution (introduce a field on a class, and add it to the constructor, for getting DI work. :(
@prochan2 mentiond to use Metalama SDK, but this option seems too complicated to me.

**Reply by gfraiteur** [reply]
Indeed you should not use the SDK. It's highly complex and poorly integrated so it should not be recommended for normal use cases.
